Fixed #29280 -- Add support for `transaction_mode` in OPTIONS

anze3db opened the pull request:
Opening this as a draft pull request until I figure out the following:

- [ ] Where to add validation for the transaction_mode values?

Right now it's in `get_connection_params`, but it doesn't feel like the right place. We do validation for `NAME` there though, so maybe it's ok? 

- [ ] Is `OPTIONS` the right place for this `transaction_mode`?

Just double checking, since it doesn't feel clean that I have to pop it from the `OPTIONS` dict to avoid `TypeError: 'transaction_mode' is an invalid keyword argument for Connection()`

- [ ] Figure out which documentation files to update

charettes commented on django/db/backends/sqlite3/base.py:
```diff
@@ -170,6 +170,18 @@ def get_connection_params(self):
+            "IMMEDIATE",
+            "EXCLUSIVE",
+        ):
+            raise ImproperlyConfigured(
+                "settings.DATABASES is improperly configured. "
+                "OPTIONS['transaction_mode'] value is invalid."
+            )
```
You might want to include the value of `transaction_mode` in the error message to help diagnose 

```suggestion
            raise ImproperlyConfigured(
                f"settings.DATABASES[{self.alias!r}]['OPTIONS']['transaction_mode'] is improperly configured. "
                f"{transaction_mode} must be one of {self.allowed_transaction_modes!r}."
            )
```

anze3db replied:
Great idea 👍

charettes commented on tests/backends/sqlite/tests.py:
```diff
@@ -245,3 +246,55 @@ def create_object():
+
+
+@unittest.skipUnless(connection.vendor == "sqlite", "SQLite tests")
+@override_settings(DEBUG=True)
```
Avoid doing that, use `CaptureQueriesContext` instead.

charettes commented on tests/backends/sqlite/tests.py:
```diff
@@ -245,3 +246,55 @@ def create_object():
+            pass
+        self.assertEqual(connection.queries[0]["sql"], "BEGIN")
+
+    def test_invalid_transaction_mode(self):
+        connection = self._get_connection_for_transaction_mode("invalid")
+        with self.assertRaises(ImproperlyConfigured):
+            connection.ensure_connection()
```
Use `override_settings` here instead and assert against the message as well using `assertRaisesMessage`

anze3db replied:
I did try to use override_settings initially but it didn't work. I spend some more time on it trying to figure out why, but I think it's because transaction test case itself uses the DATABASES setting to bootstrap itself up.

charettes replied:
That's the case but only for the `DATABASE['default']` entry by default. Try doing something like

```python
with self.modify_settings(DATABASES=settings.DATABASES | settings_dict)
``` 

by making sure to use a different alias than `"default"` and returning `connections[test_alias]` instead.

That might force you to make `_get_connection_for_transaction_mode` a context manager but that could be used to close the connection immediately after the test as well.

If that doesn't work it's not the end of the world, your approach is relatively clean, but we usually try to avoid using internal APIs if possible during tests as internal restructuration then requires adjusting them and it could result in a loss of intent from the original test.

charettes replied:
Seems like postgres backend tests [use a different approach where they replace `settings_dict`](https://github.com/django/django/blob/4879907223d70ee1a82474d9286ccfa5dae96971/tests/backends/postgresql/test_creation.py#L22-L38) and rely on `SimpleTestCase` to avoid the pre-existing connection creation.

charettes replied:
Similar situation with [the mysql tests](https://github.com/django/django/blob/4879907223d70ee1a82474d9286ccfa5dae96971/tests/backends/mysql/test_creation.py#L61-L73).

So maybe the right approach here is to use `SimpleTestCase` with `databases = {"default"}` to prevent implicit connection creation, make sure to close it on `tearDown`, and get more freedom as your test methods both test failed connection creation and the generated SQL from `atomic`.

anze3db replied:
I have updated the PR based on your suggestions, I think it looks cleaner now, but let me know if we can make it even better.

charettes commented on tests/backends/sqlite/tests.py:
```diff
@@ -245,3 +246,55 @@ def create_object():
+                )
+
+    def _get_connection_for_transaction_mode(self, transaction_mode):
+        settings_dict = {
+            "default": {
+                "ENGINE": "django.db.backends.sqlite3",
+                "NAME": ":memory:",
+                "TEST": {"NAME": ":memory:"},
+                "OPTIONS": {
+                    "transaction_mode": transaction_mode,
+                },
+            }
+        }
+        connection = ConnectionHandler(settings_dict)["default"]
+        self.addCleanup(connection.close)
```
I'd suggest `overried_settings` and accessing `connections[alias_created_for_tests]` instead of dealing with the low level connection handler machinery.

charettes reviewed (commented):
Thanks for the PR @anze3db

> Right now it's in get_connection_params, but it doesn't feel like the right place. We do validation for NAME there though, so maybe it's ok?

I think it's the right place. Look at [how we do things for postgres for example](https://github.com/django/django/blob/4879907223d70ee1a82474d9286ccfa5dae96971/django/db/backends/postgresql/base.py#L190-L250)

> Just double checking, since it doesn't feel clean that I have to pop it from the OPTIONS dict to avoid TypeError: 'transaction_mode' is an invalid keyword argument for Connection()

That's how we do it for other backend we support non-client specific options for. e.g. the postgres backend's `isolation_level` and `assume_role`.

felixxm commented on django/db/backends/sqlite3/base.py:
```diff
@@ -297,7 +314,11 @@ def _start_transaction_under_autocommit(self):
+        if transaction_mode is None:
+            self.cursor().execute("BEGIN")
+        else:
+            self.cursor().execute(f"BEGIN {transaction_mode.upper()}")
```
We don't allow calling methods in f-strings. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

felixxm commented on tests/backends/sqlite/tests.py:
```diff
@@ -245,3 +247,56 @@ def create_object():
+    def test_invalid_transaction_mode(self):
+        from django.db.backends.sqlite3.base import DatabaseWrapper
+
+        with self.change_transaction_mode("invalid"):
+            msg = (
+                "settings.DATABASES['default']['OPTIONS']['transaction_mode'] "
+                "is improperly configured to `invalid`. It must be one of "
+                "('DEFERRED', 'IMMEDIATE', 'EXCLUSIVE')."
+            )
```
`msg` can be defined outside of a context processor:
```python
msg = ...
with self.change_transaction_mode("invalid"):
```

anze3db commented:
Thank you for the early reviews @felixxm and @charettes. I believe I have addressed your feedback. I have also pushed the documentation changes so I'd appreciated feedback on that as well.

felixxm commented on docs/ref/databases.txt:
```diff
@@ -870,6 +870,26 @@ If you're getting this error, you can solve it by:
+Setting ``transaction_mode``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You might see the "Database is Locked" errors before
+the specified timeout when working with transactions. To resolve this, change 
+the transaction mode to ``IMMEDIATE``::
```
This doesn't explain anything to me as a user, check out other [docs](https://docs.djangoproject.com/en/5.0/ref/databases/#mysql-isolation-level) on inspiration what to say.

anze3db replied:
I've expanded the whole doc with a lot more information. I'm struggling because the problem that IMMEDIATE solves only happens under a very particular set of conditions, but it's very common when you have writes from multiple connecitons. 

I think ideally, Django users shouldn't have to care/read about this and Django should just do the right thing. With SQLite this means enabling WAL and using IMMEDIATE. Maybe once this PR and https://github.com/django/django/pull/14824 are merged we could update the startproject template with these settings so that new projects have a good set of defaults when starting out.

charettes replied:
Whether it's the _right thing_ to do by default is a whole can of worms that we cannot assume here.

We haven't even run the whole test suite with these options enabled to confirm it doesn't break anything, setup CI to continiously do so, and confirm there is an upgrade path for the 10+ years old projects out there using SQLite if any problem is discovered.

It's one thing to provide an option, which this PR and the associated docs should focus on, and a completely different one to make it the default, which we can't assume here.

anze3db replied:
Yes, sorry if I was unclear. I'm not proposing to change the default. I don't want to change how 10+ year old projects are configured.

I only wanted to mention that in the future, we could discuss enabling IMMEDIATE + WAL for **new** projects (the settings.py file generated for new projects would have these settings) - so that they have an SQLite config that's optimized for multiple processes/threads and not an embedded system.

Sorry again for derailing the conversation.

felixxm reviewed (commented):
@anze3db Thanks :+1: I left more comments.

charettes reviewed (commented):
I left comments on the proposed docs but otherwise the patch LGTM

felixxm commented on docs/ref/databases.txt:
```diff
@@ -870,6 +870,52 @@ If you're getting this error, you can solve it by:
+transaction to a write transaction. If some other connection modified the
+database during this, upgrading to a write transaction is impossible, and the
+"Database is Locked" error is raised immediately, regardless of the
+``timeout`` setting.
```
We should not copy SQLite docs here.

felixxm reviewed (commented):
@anze3db Thanks :+1: I pushed edits. Let me know what you think.

anze3db reviewed (commented):
Looks good to me 👍 thank you @felixxm!

felixxm commented:
@anze3db Welcome aboard :boat:

anze3db commented:
Awesome! Thank you @felixxm!
